component/encryptor/encryptor.go

```diff
@@ -0,0 +1,368 @@
+		if os.IsNotExist(err) {
+			log.Trace("Encryptor::GetAttr : File not found: %s", options.Name)
+			return nil, syscall.ENOENT
+		}
+		log.Trace("Encryptor::GetAttr : Error getting file attributes: %s", err.Error())
+		return &internal.ObjAttr{}, err
+	}
+
+	// Populate the ObjAttr struct with the file info.
+	attr = &internal.ObjAttr{
+		Mtime:  fileAttr.ModTime(),                // Modified time
+		Atime:  time.Now(),                        // Access time (current time as approximation)
+		Ctime:  fileAttr.ModTime(),                // Change time (same as modified time in this case)
+		Crtime: fileAttr.ModTime(),                // Creation time (not available in Go, using modified time)
+		Mode:   fileAttr.Mode(),                   // Permissions
+		Path:   e.mountPointCipher + options.Name, // Full path
+		Name:   fileAttr.Name(),                   // Base name of the path
+	}
+	fileHandle, err := os.OpenFile(e.mountPointCipher+options.Name, os.O_RDONLY, 0666)
+	if err != nil {
+		log.Trace("Encryptor::GetAttr : Error opening file: %s", err.Error())
+		return nil, err
+	}
+	defer fileHandle.Close()
+
+	if fileAttr.IsDir() {
+		attr.Size = fileAttr.Size()
+		attr.Flags.Set(internal.PropFlagIsDir)
+	} else {
+		actualFileSize, err := checkForActualFileSize(fileHandle, fileAttr.Size(), int64(e.blockSize))
+		if err != nil {
+			log.Err("Encryptor: Error checking for actual file size: %s", err.Error())
+			return nil, err
+		}
+		attr.Size = actualFileSize
+	}
+	return attr, nil
+}
+
+// Encrypt the incoming block of data using the encryption key and writing it
+// to the cipher mount point.
+func (e *Encryptor) StageData(opt internal.StageDataOptions) error {
+	log.Info("Encryptor::StageData : %s, offset %d, data %d", opt.Name, opt.Offset, len(opt.Data))
+
+	paddingLength := int64(0)
+	dataLen := int64(len(opt.Data))
+	blockId := opt.Offset
+	if dataLen < int64(e.blockSize) {
+		// Pad the data to the block size.
+		paddingLength = int64(e.blockSize) - dataLen
+		opt.Data = append(opt.Data, make([]byte, paddingLength)...)
+	}
+	encryptedChunk, nonce, err := EncryptChunk(opt.Data, e.encryptionKey)
+	if err != nil {
+		log.Err("Encryptor: Error encrypting data: %s", err.Error())
+		return err
+	}
+
+	encryptedChunkOffset := int64(blockId) * (int64(e.blockSize) + int64(MetaSize))
+	// Write the combined nonce and encrypted chunk.
+	n, err := e.handle.WriteAt(append(nonce, encryptedChunk...), encryptedChunkOffset)
+	log.Debug("Encryptor::StageData : writing %d bytes to encrypted file: %s at offset %d", n, opt.Name, encryptedChunkOffset)
+	if err != nil {
+		log.Err("Encryptor: Error writing encrypted chunk to encrypted file: %s at offset %d : size of data %d", err.Error(), encryptedChunkOffset, n)
+		return err
+	}
+
+	e.lastChunkMeta.Lock()
+	defer e.lastChunkMeta.Unlock()
+	if int64(blockId) > e.lastChunkMeta.farthestBlockSeen {
+		e.lastChunkMeta.farthestBlockSeen = int64(blockId)
+	}
+
+	e.lastChunkMeta.paddingLength = paddingLength
```

Comment:
Shouldn't you be updating the paddingLength also inside the if block ?

Reply:
kept it outside to take the default value(0) of paddingLength in case it doesn't go in if block.